src/common/Gateway.sol

```diff
@@ -271,47 +276,46 @@ contract Gateway is Auth, IGateway, Recoverable {
         IAdapter[] memory adapters_ = adapters[centrifugeId];
         require(adapters[centrifugeId].length != 0, EmptyAdapterSet());
 
-        uint64 batchGasLimit_ =
-            (isBatching) ? batchGasLimit[centrifugeId][poolId] : gasService.gasLimit(centrifugeId, batch);
-        uint64 proofGasLimit = gasService.gasLimit(centrifugeId, proof);
+        uint64 gasLimit = (isBatching) ? batchGasLimit[centrifugeId][poolId] : messageGasLimit;
 
         for (uint256 i; i < adapters_.length; i++) {
-            IAdapter currentAdapter = IAdapter(adapters_[i]);
             bool isPrimaryAdapter = i == PRIMARY_ADAPTER_ID - 1;
-            bytes memory payload = isPrimaryAdapter ? batch : proof;
-            uint64 gasLimit = isPrimaryAdapter ? batchGasLimit_ : proofGasLimit;
 
-            uint256 consumed = currentAdapter.estimate(centrifugeId, payload, gasLimit);
+            uint256 consumed = adapters_[i].estimate(centrifugeId, isPrimaryAdapter ? batch : proof, gasLimit);
```

Comment:
I needed to play here to avoid stack-deep issues

Reply:
This will add another storage lookup. Wonder if we can accomplish this another way 🤔

Reply:
`adapters_` is already in memory, though. So I think it should not be from the storage. I tried everything there and that was the only way I found.

Reply:
That's true, then it seems fine!